GSOC'25 : Set up sequential solver - solver-portfolio

RehanChalana opened the pull request.

RehanChalana committed:
feat : set up sequential solver

- refactor dp configuration from a String Array to Set of strings

- refactor SymbolicConstraintsGeneral.isSatisfiable() to allow multiple solvers to try and solve a PC

- throw NoSolverSucceededException if No solver is able to solve or parse a PC

# Conflicts:
#	jpf-symbc/src/main/gov/nasa/jpf/symbc/SymbolicInstructionFactory.java

RehanChalana committed:
fix : skip if choco returns UNSAT or false

RehanChalana committed:
fix : NullPointerException in cleanup() with a null check

RehanChalana committed:
fix : sequential solver fixes

- convert dp from a Set<String> to List<String> to preserve order

- return a false in isSatisfiable if the PCParser.parse returns null

- Uncomment "omega" hacks

RehanChalana committed:
chore : update jpf-sv-comp script

RehanChalana committed:
fix: change java path back

sohah commented on jpf-sv-comp:
```diff
@@ -41,13 +41,17 @@ mkdir -p $DIR/target/classes
 echo "symbolic.arrays=true" >> $DIR/config.jpf
+echo "listener=.symbc.SymbolicListener" >> $DIR/config.jpf
+echo "symbolic.min_int=-21474836" >> $DIR/config.jpf
+echo "symbolic.max_int=21474836" >> $DIR/config.jpf
```
are these for choco, right? z3 should be able to support a bigger range. I suggest we have these as the normal range, but if choco solver is used we bring that range down, while printing a WARNING to users, that is clear that choco is working with a smaller range.
In fact, I think if the user uses choco with or without sequential solving, we should print out a warning that choco is not sound for verification, i.e., saying unsat, because it works with a smaller int/double/float ranges.

RehanChalana replied:
Yes these are the ranges for choco, as suggested I have removed these two lines and hence default ranges will be used which are Integer.MIN_VALUE and Integer.MAX_VALUE and if we choco is used as solver a warning will be printed informing the users and instead of crashing we will reduce the ranges to choco's safe ranges.

Also if user chooses to use more solvers with choco we won't trust choco's UNSAT/False but if only choco is selected then I think we should allow it as we are already printing a warning to the users.

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNE.java:
```diff
@@ -47,14 +49,15 @@ public Instruction execute(ThreadInfo ti) {
                 if (SymbolicInstructionFactory.collect_constraints)
                     cg = new PCChoiceGenerator(1);
-                else if (dp[0].equalsIgnoreCase("omega")) // hack because omega does not handle not or or correctly
+//                TODO: do we support omega ?
```
Can we remove these comments. And please make sure to write in your documentation about omega, that we further need to investigate that.

RehanChalana replied:
done

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNE.java:
```diff
@@ -93,7 +96,8 @@ else if (dp[0].equalsIgnoreCase("omega")) // hack because omega does not handle
 
             if (conditionValue) {
-                if (dp[0].equalsIgnoreCase("omega")) {// hack
+//                TODO: do we still support omega solver ?
```
same comment about omega, please address everywhere

RehanChalana replied:
done

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/NoSolverSucceededException.java:
```diff
@@ -0,0 +1,7 @@
+package gov.nasa.jpf.symbc.numeric;
```
any reason why you have this exception in the numeric package?

RehanChalana replied:
Yes, Because all the related classes were in the numeric package as well. For ex all the ProblemGeneral Impls, PathCondition, SymbolicConstraintsGeneral all are in the numeric package as well.

sohah replied:
@RehanChalana , well, this is not general enough. I do not see an SPF exception. So, I think we need one in the outer most package with a general name like "SPFException" then within the message you can have "No solver succeeded" message.

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/SymbolicConstraintsGeneral.java:
```diff
@@ -46,13 +46,17 @@
 public class SymbolicConstraintsGeneral {
-    protected ProblemGeneral pb;
+    protected List<ProblemGeneral> solvers;
+    protected ProblemGeneral resultSolver;
```
Can you please add comments about these two variables? please use the java doc formating.

RehanChalana replied:
done

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/SymbolicInstructionFactory.java:
```diff
@@ -538,7 +543,7 @@ public Instruction multianewarray(String clsName, int dimensions){
 	      }
 
-	static public String[] dp;
+	static public List<String> dp;
```
Can you please write a comment about this variable?

RehanChalana replied:
done

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/SymbolicConstraintsGeneral.java:
```diff
@@ -76,71 +80,100 @@ public boolean isSatisfiable(PathCondition pc) {
+            try {
+
+                ProblemGeneral tempPb = PCParser.parse(pc, resultSolver);
+
```
why are you leaving a lot of space?

RehanChalana replied:
fixed

sohah reviewed (changes requested):
@RehanChalana , can you please address the few comments. Thanks!

RehanChalana committed:
fix : improve handling of choco's safe range

- use z3 as the default solver when symbolic.dp config is not present

_ log a warning when choco is selected in the symbolic.dp config

- instead of crashing bring down the ranges when choco is used with exceeding safe ranges

- use the correct range in jpf-sv-comp script

RehanChalana commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/SymbolicConstraintsGeneral.java:
```diff
@@ -76,71 +80,100 @@ public boolean isSatisfiable(PathCondition pc) {
-            throw new RuntimeException(
-                    "## Error: unknown decision procedure symbolic.dp=" + dp[0] + "\n(use choco or IAsolver or CVC3)");
-
-        /*
-         * Parse path condition to solver. Note: do not override the actual pb
-         * variable in case the result is null. The cleanup afterwards will not
-         * work otherwise and the solver gets filled up with wrong assertions,
-         * e.g. with Z3.
-         */
```
The reason I removed this comment is because we don’t actually need to override the solver.
`PCParser.parse(pc, solver) `returns the same solver instance if parsing is successful, and null if it fails.
So instead of overriding the solver, we can just store the result of `PCParser.parse()` in a temp variable and check for null.

RehanChalana committed:
doc : improve comments and formatting

RehanChalana commented:
@sohah I have addressed all of your comments and updated the pr as well, Can you please take a look again.

Thanks!

RehanChalana committed:
Merge branch 'solver-portfolio' into solver-portfolio

RehanChalana committed:
fix : import java.util.List in SymbolicListener

sohah commented on jpf-sv-comp:
```diff
@@ -41,7 +41,7 @@ mkdir -p $DIR/target/classes
 # create configuration file
 echo "target=Main" > $DIR/config.jpf
 echo "classpath=`pwd`/jpf-symbc/build/classes:$DIR/target/classes:$CLASSPATH_ADDED" >> $DIR/config.jpf
-echo "symbolic.dp=z3" >> $DIR/config.jpf
+echo "symbolic.dp=z3,choco" >> $DIR/config.jpf
 echo "symbolic.bvlength=64" >> $DIR/config.jpf
 echo "search.depth_limit=13" >> $DIR/config.jpf
 echo "symbolic.strings=true" >> $DIR/config.jpf
```
In addition to have this. I think we need a test case that is seperate that is just testing sequential solving. Can you please add a configuration file for this, along with any java example that is a good candidate for testing?

sohah reviewed (changes requested):
@RehanChalana , a couple of more comments. Can you please address them and let me know? Thank you!

RehanChalana committed:
refactor : change NoSolverSucceededException.java with SPFException and a message

- add SequentialTest1 example to depict the sequential behaviour

RehanChalana commented:
Hi @sohah 

I have addressed your comments and I have created an SPFException and I am using this instead of NoSolverSucceddedException  and I have also added a test case in a separate "sequential" directory, I have enabled the debug option in the jpf file so that the sequential behavior is clearly visible.

Can you please take a look again?

Thanks!

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/SymbolicInstructionFactory.java:
```diff
@@ -661,23 +669,56 @@ public  SymbolicInstructionFactory (Config conf){
+						"WARNING: Choco solver is enabled.%n" +
+								"   - Choco works with reduced integer ranges: [%d, %d].%n" +
+								"   - UNSAT results may be UNSOUND for verification.%n" +
+								"   - Use with caution, especially in verification tasks.%n",
```
can we remove this line. I think it is redundent

RehanChalana replied:
done

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/SymbolicConstraintsGeneral.java:
```diff
@@ -76,71 +92,91 @@ public boolean isSatisfiable(PathCondition pc) {
+                    // skip if choco returns an UNSAT/false if more than one solver is used
+                    if (solvers.size() > 1 && resultSolver instanceof ProblemChoco && result == false) {
+                        result = null;
+                        continue;
```
Please do not use continue and break within a `for` loop. Remember you can have conditions next to `i++`, and you an also have an if-then-else statement that would allow you to `continue`.
Or the other option is to use a `while` loop here, but again, please no `break` and no `continue`.

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/SymbolicConstraintsGeneral.java:
```diff
@@ -76,71 +92,91 @@ public boolean isSatisfiable(PathCondition pc) {
-                    ((ProblemZ3Optimize) pb).maximize(
-                            PCParser.getExpression((IntegerExpression) Observations.lastObservedSymbolicExpression));
+
+        for(int i = 0; i < solvers.size(); i++) {
```
Can we move this to a seperate method? and can we have a Java style documentation on what it is doing, and also please add an explaination as to why do we skip when choco gives us unsat?

RehanChalana replied:
do you mean moving the whole loop into a separate function ?

sohah replied:
Yes, the currently `for` loop should be in a seperate method for readability

RehanChalana replied:
done

sohah reviewed (changes requested):
@RehanChalana , a couple of more comments and I think we can accept this PR after them.

Thank you!

RehanChalana committed:
refactor: move the sequential solving loop into a separate method for better readability

RehanChalana commented:
I have addressed the changes @sohah 

Can you please review again

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/SymbolicConstraintsGeneral.java:
```diff
@@ -164,6 +162,54 @@ else if (dp[0].equalsIgnoreCase("no_solver")) {
+     *
+     * @param pc       the PathCondition to solve
+     **/
+    public void checkPathConditionSequentially(PathCondition pc) {
```
why is this returning a void, not a Boolean?

RehanChalana replied:
because instead of returning a Boolean, it sets the class field "result" which is also a Boolean. I was thinking of making it return a Boolean as well. But we also need this function to tell and return the solver instance that solved the PC. 

So either we create some sort of Pair<Boolean, ProblemGeneral> object that can be returned from this method? Or is the current implementation is fine ? where we set the class fields instead of returning.

sohah replied:
@RehanChalana , let's be explicit about what this method do, and that is usually useful if we return a specific objects, instead of having implicit side effects. We can do the latter but we have to be very clear that we are doing this.

In this case, I'd rather return `int[]` that will carry two objects. If you use `Pair`, I think you'll have to add a new package to SPF, I am not sure, you can double check me on that, if the package is already imported, then we can use `Pair`

sohah replied:
@RehanChalana , to clarify, I am talking about whether the package that is including `Pair` is imported to SPF project, i.e., is included as a dependency in gradle.

RehanChalana replied:
We do have already have the dependency for `Pair`

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/SymbolicConstraintsGeneral.java:
```diff
@@ -164,6 +162,54 @@ else if (dp[0].equalsIgnoreCase("no_solver")) {
+     * It stops as soon as one solver can parse and return a definite SAT/UNSAT result.
+     *
+     * @param pc       the PathCondition to solve
+     **/
```
This is incomplete description because we are not mentioning the exceptional case of choco and its unsoundness. It sounds trivial, but it is very important to be clear about what is going on here.

sohah reviewed (changes requested):
@RehanChalana , can you please address this additional comments? Thank you!

RehanChalana commented:
@sohah Can you check again?

I have improved the comment and as well as returning a Pair<Boolean, ProblemGeneral> in the method as we discussed!

sohah reviewed (approved):
@RehanChalana , thank you!
@yannicnoller , can we please accept this PR. I do not see the "resolve comments" botton that you were mentioning the other day. I think it might privilege related. However, I checked all the comments and they are all addressed. Thank you!
